### DISCUSSION TITLE
Native Image with Hibernate - Missing Hibernate GraalVM Feature?

### DISCUSSION THREAD

**Discussion by saw303** [micronaut-projects/micronaut-core#8071] [Q&A] [answered]
Hi,

I'm currently trying to get a rather mature Micronaut application with plenty of JPA entities to run as a native image. It seems like that the provided `io.micronaut.configuration.hibernate.jpa.graal.HibernateFeature` simply provides basic stuff around JDBC drivers and Hibernate dialects.

I therefore implemented my own GraalVM feature (code is still a prototype). Beside of projects specific stuff (NamingStrategies, CustomUserTypes and Embeddables, I had to define plenty of Hibernate specific classes since heavy reflection is involved.

```java
@AutomaticFeature
final class SalesServerFeature implements Feature {

  @Override
  public void beforeAnalysis(BeforeAnalysisAccess access) {

    // I had to register project specific NamingStrategies and custom Hibernate user types.
   // I had to register project specific embeddables such as
    AutomaticFeatureUtils.registerConstructorsForRuntimeReflection(
        access, "ch.onstructive.salessystem.embeddables.OrganisationRef");

   // Hibernate related stuff.
    AutomaticFeatureUtils.registerConstructorsForRuntimeReflection(
        access, "org.hibernate.persister.entity.JoinedSubclassEntityPersister");
    AutomaticFeatureUtils.registerConstructorsForRuntimeReflection(
        access, "org.hibernate.persister.entity.SingleTableEntityPersister");
    AutomaticFeatureUtils.registerConstructorsForRuntimeReflection(
        access, "org.hibernate.persister.entity.UnionSubclassEntityPersister");
    AutomaticFeatureUtils.registerConstructorsForRuntimeReflection(
        access, "org.hibernate.persister.collection.OneToManyPersister");
    AutomaticFeatureUtils.registerConstructorsForRuntimeReflection(
        access, "org.hibernate.persister.collection.BasicCollectionPersister");

    AutomaticFeatureUtils.registerConstructorsForRuntimeReflection(
        access, "org.hibernate.sql.ordering.antlr.NodeSupport");
    AutomaticFeatureUtils.registerConstructorsForRuntimeReflection(
        access, "org.hibernate.sql.ordering.antlr.SortKey");
    AutomaticFeatureUtils.registerConstructorsForRuntimeReflection(
        access, "org.hibernate.sql.ordering.antlr.OrderingSpecification");
    AutomaticFeatureUtils.registerConstructorsForRuntimeReflection(
        access, "org.hibernate.sql.ordering.antlr.SortSpecification");
    AutomaticFeatureUtils.registerClassForRuntimeReflectionAndReflectiveInstantiation(
        access, "org.hibernate.sql.ordering.antlr.OrderByFragment");
    AutomaticFeatureUtils.registerConstructorsForRuntimeReflection(access, "antlr.CommonToken");
  }
}
```

My question is, wouldn't it make sense to put the Hibernate related code into the existing Micronaut `HibernateFeature` or should this be somewhere else?

Or am I doing something wrong?

Furthermore I have noticed that the `FlywayFeature` only works, when the migration scripts are located in the default location (see https://github.com/micronaut-projects/micronaut-flyway/blob/2a4f58af64787ed1ff0c9bdbefbebba3464569e8/flyway/src/main/java/io/micronaut/flyway/graalvm/FlywayFeature.java#L60). Whats the idea here, when the location has been changed in the `application.yml`? Does a feature have access to this configuration or does it currently require to either use the default location or implement a project specific feature?

**Comment by graemerocher** [comment, marked as answer]
PRs welcome for the hibernate specific code to the existing HibernateFeature

As for Flyway you could I guess load the environment within the flyway feature and try read the configuration location, would be a separate improvement.

**Reply by saw303** [reply]
Hibernate PR filed at https://github.com/micronaut-projects/micronaut-sql/pull/727
